pkg/liqo-controller-manager/external-network/wireguard/wggatewayserver_controller.go

```diff
@@ -0,0 +1,370 @@
+// Copyright 2019-2023 The Liqo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wireguard
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"golang.org/x/exp/maps"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	networkingv1alpha1 "github.com/liqotech/liqo/apis/networking/v1alpha1"
+	"github.com/liqotech/liqo/pkg/consts"
+	"github.com/liqotech/liqo/pkg/discovery"
+	"github.com/liqotech/liqo/pkg/utils"
+)
+
+// WgGatewayServerReconciler manage WgGatewayServer lifecycle.
+type WgGatewayServerReconciler struct {
+	client.Client
+	Scheme           *runtime.Scheme
+	extNetPodsClient client.Client
+}
+
+// NewWgGatewayServerReconciler returns a new WgGatewayServerReconciler.
+func NewWgGatewayServerReconciler(cl client.Client, s *runtime.Scheme, extNetPodsClient client.Client) *WgGatewayServerReconciler {
+	return &WgGatewayServerReconciler{
+		Client:           cl,
+		Scheme:           s,
+		extNetPodsClient: extNetPodsClient,
+	}
+}
+
+// +kubebuilder:rbac:groups=networking.liqo.io,resources=wggatewayservers,verbs=get;list;watch;delete;create;update;patch
+// +kubebuilder:rbac:groups=networking.liqo.io,resources=wggatewayservers/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;delete;create;update;patch
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;delete;create;update;patch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch
+
+// Reconcile manage GatewayServer lifecycle.
+func (r *WgGatewayServerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	wgServer := &networkingv1alpha1.WgGatewayServer{}
+	if err = r.Get(ctx, req.NamespacedName, wgServer); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof("WireGuard gateway server %q not found", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf("Unable to get the WireGuard gateway server %q: %v", req.NamespacedName, err)
+		return ctrl.Result{}, err
+	}
+
+	requeue := false
```

Comment:
this is unnecessary, it should be handled by the Owns functions triggering other reconciliations